### PR TITLE
WIP: Modal component

### PR DESCRIPTION
**Pull request by rauldudu**
### 🛠 Purpose

- [x] Migrate same code from acl-ui
- [x] Use style components
- [ ] Refactor code to use hooks
- [ ] Add tests
- [ ] Add documentation

### ✏️ Notes to Reviewer

---

### 🖥 Screenshots

---

### CODE REVIEW

File: packages/Modal/package.json [hunk 1]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "@paprika/modal",
+  "version": "0.0.1",
```
Comment (alexzherdev) [inline packages/Modal/package.json]:
```suggestion
  "version": "0.1.0",
```

File: packages/Modal/src/Modal.js [hunk 2]
```diff
@@ -0,0 +1,149 @@
+/* eslint-disable jsx-a11y/no-static-element-interactions */
+
+import React from "react";
+import ReactDOM from "react-dom";
+import PropTypes from "prop-types";
+import classNames from "classnames";
```
Comment (alexzherdev) [inline packages/Modal/src/Modal.js]: missing this in package.json

File: packages/Modal/src/Header/Header.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    return (
+      <header className={className} css={headerStyles} isDark={isDark}>
+        {this.renderCloseButton()}
+        <div css={headingStyle} isDark={isDark} role="heading" aria-level={headerAriaLevel}>
+          {children}
+        </div>
+      </header>
+    );
+  }
+}
```
Comment (AndreyChernykh) [inline packages/Modal/src/Header/Header.js]: I'd rewrite this component as 

```js
function Header({ hasCloseButton, isDark, onClose }) {
     return (
      <header className={className} css={headerStyles} isDark={isDark}>
        {hasCloseButton ? <Button.Close isCloseButton size="small" isDark={isDark} cssOverrides={closeButtonStyle} onClick={onClose} /> : null}
        <div css={headingStyle} isDark={isDark} role="heading" aria-level={headerAriaLevel}>
          {children}
        </div>
      </header>
    );
  }
}
```

File: packages/Modal/src/Modal.js [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+  width: "medium",
+  zIndex: null,
+};
+
+let modalRoot;
+
+class Modal extends React.Component {
+  constructor(props) {
+    super(props);
+    this.modalContainer = null;
```
Comment (AndreyChernykh) [inline packages/Modal/src/Modal.js]: is not used

File: packages/Modal/src/Modal.js [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    this.el = document.createElement("div");
+  }
+
+  componentDidMount() {
+    document.addEventListener("keydown", this.handleKeyDown, false);
+    modalRoot.appendChild(this.el);
+  }
+
+  componentDidUpdate() {
+    this.renderModal();
```
Comment (AndreyChernykh) [inline packages/Modal/src/Modal.js]: How does it even work? 🤔

Comment (alexzherdev) [reply packages/Modal/src/Modal.js]: <magic.gif>

File: packages/Modal/src/Modal.js [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  render() {
+    return ReactDOM.createPortal(this.renderModal(), this.el);
+  }
+}
+
+Modal.displayName = "Modal";
+Modal.propTypes = propTypes;
+Modal.defaultProps = defaultProps;
+
+export default Modal;
```
Comment (AndreyChernykh) [inline packages/Modal/src/Modal.js]: Why don't we use some ↪️ in here? 😏 

(may contain some bugs) 
```js
function useEscHandler(handleKeyDown) {
  React.useEffect(() => {
    function handleKeyDown(event) {
      if (event.key === "Escape") {
        handleKeyDown();
      }
    }

    document.addEventListener("keydown", handleKeyDown, false);

    return function cleanup() {
      document.removeEventListener("keydown", handleKeyDown);
    };
  }, [handleKeyDown]);
}

function useModalDiv() {
  const modalRoot = React.useRef(null);
  const el = React.useRef(null);

  React.useEffect(() => {
    modalRoot.current = document.createElement("div");
    document.body.appendChild(modalRoot.current);

    el.current = document.createElement("div");
    modalRoot.current.appendChild(el.current);

    return function cleanup() {
      modalRoot.current.removeChild(el.current);
      // Not sure about modal root, though. Why do we need it even? Shouldn't we clean it up?
    };
  }, []);

  return el.current;
}

function Modal(props) {
  const ariaId = React.useRef(uuid());
  const $el = useModalDiv();
  const portalRef = React.ureRef();

  const { onEscKey } = props;

  const handleKeyDown = React.useCallback(
    event => {
      if (event.key === "Escape") {
        onEscKey();
      }
    },
    [onEscKey]
  );

  useEscHandler(handleKeyDown);

  const setRef = React.useCallback(
    node => {
      if (node !== null) {
        portalRef.current = node.getBoundingClientRect;
        this.props.portalRef(node);
      }
    },
    [portalRef]
  );

  return ReactDOM.createPortal(
    () => (
      // eslint-disable-next-line jsx-a11y/no-static-element-interactions
      <div
        className={classNames(className, { "modal--is-open": props.isOpen })}
        css={modalStyles}
        isOpen={props.isOpen}
        width={props.width}
        onKeyDown={handleKeyDown}
        ref={setRef}
        zIndex={props.zIndex}
      >
        {props.hasOverlay ? (
          <div
            css={props.overlayStyles}
            onClick={props.onOverlayClick}
            onKeyDown={props.onOverlayClick}
            tabIndex="-1"
          />
        ) : null}
        <div css={frameStyles} width={props.width}>
          <section
            css={contentStyles}
            isOpen={props.isOpen}
            isScrollable={props.isScrollable}
            role="dialog"
            aria-labelledby={this.ariaId}
          >
            <span css={stylers.visuallyHidden} id={this.ariaId}>
              {props.ariaLabel}
            </span>
            {props.children}
          </section>
        </div>
      </div>
    ),
    $el
  );
}

```

Comment (AndreyChernykh) [reply packages/Modal/src/Modal.js]: @nahumzs @alexzherdev @rauldudu @mikrotron

File: packages/Modal/src/Modal.js (hunk 5 above)
Comment (rauldudu) [reply packages/Modal/src/Modal.js]: Screen Shot 2019-07-19 at 10 59 24 AM
@mikrotron may have the answer :P

File: packages/Modal/src/Modal.js (hunk 6 above)
Comment (rauldudu) [reply packages/Modal/src/Modal.js]: Just to give some context, the JS part is basically a copy/paste from acl-ui. Next step would be refactoring using hooks...

Comment (mikrotron) [reply packages/Modal/src/Modal.js]: Yep, save that snippet for round 2! Thanks, @AndreyChernykh!

File: packages/Modal/src/Modal.js (hunk 5 above)
Comment (mikrotron) [reply packages/Modal/src/Modal.js]: Naîveté? It doesn't actually depend on this to work, does it? Can you just remove it?

Comment (alexzherdev) [reply packages/Modal/src/Modal.js]: This will get refactored though anyway?

Comment (rauldudu) [reply packages/Modal/src/Modal.js]: Right it will end up being removed

File: packages/Modal/src/Modal.js (hunk 6 above)
Comment (AndreyChernykh) [reply packages/Modal/src/Modal.js]: @mikrotron You're welcome!

Comment (nahumzs) [reply packages/Modal/src/Modal.js]: I just saw this, but is a great idea man!!

**Comment by nahumzs** [comment]
`fyi` I'm closing this one for now, and I will let @DanilAgafonov decide later if he wants to reopen this PR or create a new one for the Modal when he gets to it.

**Pull request closed without merging**